include/dlaf/eigensolver/tridiag_solver/rot.h

```diff
@@ -0,0 +1,175 @@
+  // TODO allocate just sub-range
+  matrix::Panel<Coord::Col, T, D> workspace(dist_ws);
+
+  auto givens_rots_fn = [comm_row, dist_sub, mb](std::vector<GivensRotation<T>> rots,
+                                                 std::vector<matrix::Tile<T, D>> tiles,
+                                                 matrix::Tile<T, D> tile_ws) {
+    const SizeType m = dist_sub.localSize().rows();
+
+    auto getColPtr = [&dist_sub, &tiles, &tile_ws](const SizeType col_index, const bool hasIt) -> T* {
+      // TODO document column layout assumption
+      // TODO assume column layout and operate on the full column? also workspace must obey
+      if (hasIt) {
+        const LocalTileIndex tile_col{dist_sub.nextLocalTileFromGlobalElement<Coord::Row>(0),
+                                      dist_sub.nextLocalTileFromGlobalElement<Coord::Col>(col_index)};
+        const std::size_t linear_tile_col = to_sizet(dist_sub.localTileLinearIndex(tile_col));
+        return tiles[linear_tile_col].ptr(dist_sub.tileElementIndex({0, col_index}));
+      }
+      return tile_ws.ptr({0, 0});
+    };
+
+    ex::unique_any_sender<> serializer = ex::just();
+    for (const GivensRotation<T>& rot : rots) {
+      const SizeType j_x = rot.i / mb;
+      const SizeType j_y = rot.j / mb;
+
+      const comm::IndexT_MPI rankColX = dist_sub.template rankGlobalTile<Coord::Col>(j_x);
+      const comm::IndexT_MPI rankColY = dist_sub.template rankGlobalTile<Coord::Col>(j_y);
+
+      const bool hasX = dist_sub.rankIndex().col() == rankColX;
+      const bool hasY = dist_sub.rankIndex().col() == rankColY;
+
+      if (!hasX && !hasY)
+        continue;
+
+      T* col_x = getColPtr(rot.i, hasX);
+      T* col_y = getColPtr(rot.j, hasY);
+
+      const bool hasBothXY = hasX && hasY;
+
+      std::vector<ex::unique_any_sender<>> cps;
+      if (!hasBothXY) {
+        // TODO compute TAG
+
+        const comm::IndexT_MPI rank_partner = hasX ? rankColY : rankColX;
+
+        // TODO possible optimization, check if it is zero or not
+
+        const T* col_send = hasX ? col_x : col_y;
+        T* col_recv = hasX ? col_y : col_x;
+
+        cps.emplace_back(wrapper::scheduleSendCol<D, T>(comm_row, rank_partner, 0, col_send, m));
+        cps.emplace_back(wrapper::scheduleRecvCol<D, T>(comm_row, rank_partner, 0, col_recv, m));
+      }
+
+      // each one computes his own, but just stores either x or y
+      // (or both if are on the same rank)
+      serializer =
+          di::whenAllLift(std::move(serializer), ex::when_all_vector(std::move(cps))) |
+          di::transform(di::Policy<DefaultBackend_v<D>>(), [rot, m, col_x, col_y](auto&&... ts) {
+            if constexpr (D == Device::CPU)
+              blas::rot(m, col_x, 1, col_y, 1, rot.c, rot.s);
+            else
+              // givensRotationOnDevice(m, x, y, rot.c, rot.s, ts...);
+              DLAF_STATIC_UNIMPLEMENTED(T);
+          });
+    }
```

Comment:
My understanding here is that this is done because you only have enough workspace to store a single vector at a time. That is why the Given's rotation has to be performed just after the communication is done and no other communication can be launched to prevent overriding the workspace. Is that correct? If so, perhaps you can include more details in your comment?

Reply:
Yes, exactly. I will follow your advice and document a bit the serialisation 😉